row-level ON DELETE / UPDATE CASCADE

lhirata opened the pull request:
Closes #2015, #2185.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/2379)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/24b68dde6118ab64745070ad449d81aa044268a1/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/0d27ce0592d32309248237b9edaf3eed635f9647/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a61d28f0a7c7a58931db47079fd18112d6be6cd8/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/d12a4a9b58a3640cddf7d40419b029ad83c33f56/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7f9f98db7d4029578d75a15e8f15f86337c8694e/

BramGruneir commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[v2.0/add-constraint.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47cpjiKdJ19KMHqlXB:-L47cpjiKdJ19KMHqlXC:b-zepa3b) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L107)):*
> ```Markdown
> 
> <span class="version-tag">New in v2.0:</span> You can include a [foreign key action](foreign-key.html#foreign-key-actions) to specify what happens when a foreign key is updated or deleted.
> ```

There are two bullets here.  Did you mean to not have the first one?

---

*[v2.0/add-constraint.md, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47cx1ds0b2DGZBX4IP:-L47cx1ds0b2DGZBX4IQ:bcodqio) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L112)):*
> ```Markdown
> <span class="version-tag">New in v2.0:</span> You can include a [foreign key action](foreign-key.html#foreign-key-actions) to specify what happens when a foreign key is updated or deleted. In this example, let's use `ON DELETE CASCADE` (i.e., when referenced row is deleted, all dependent objects are also deleted).
> 
> {{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it drops, so it should be used cautiously.{{site.data.alerts.end}}
> ```

change this to "does not list the objects that it drops or updates"

---

*[v2.0/add-constraint.md, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47d6BNn7cnKfnqiPuX:-L47d6BNn7cnKfnqiPuY:b-86u82h) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L126)):*
> ```Markdown
> ## See Also
> 
> - [Constraints](constraints.html)
> ```

Add a link here to Foreign Key Actions?

---

*[v2.0/create-table.md, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47f1nYUT8fgX_Y98Hm:-L47f1nYUT8fgX_Y98Hn:b4flk4o) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/create-table.md#L329)):*
> ```Markdown
> ### Create a Table that Prevents Deletion of Rows with Dependent Objects (`RESTRICT`) <span class="version-tag">New in v2.0</span>
> 
> If a row has dependent objects, such as a foreign key reference, CockroachDB will not delete or update the row by default; however, if you want to be sure of the behavior, you can include the [foreign key action](foreign-key.html#foreign-key-actions) `ON [DELETE | UPDATE] RESTRICT` when creating a table or [adding a constraint](add-constraint.html).
> ```

I don't know if this advice is required.  By default we are NO ACTION which is a restrict right now.

---

*[v2.0/delete.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47ep3VqRNFYNcyCKZ3:-L47ep3VqRNFYNcyCKZ4:b-zayj08) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/delete.md#L9)):*
> ```Markdown
> The `DELETE` [statement](sql-statements.html) deletes rows from a table.
> 
> {{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it drops, so it should be used cautiously. If you delete a row that is referenced by a <a href="foreign-key.html">foreign key</a> and <code>ON DELETE CASCADE</code> is set, all of the dependent rows will also be deleted.{{site.data.alerts.end}}
> ```

Expand this for the other actions.

---

*[v2.0/foreign-key.md, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47dDxl7WvwSxuMF_K4:-L47dDxl7WvwSxuMF_K5:b-s56n68) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/foreign-key.md#L57)):*
> ```Markdown
> Parameter | Description
> ----------|------------
> `ON DELETE NO ACTION` / `ON UPDATE NO ACTION` | _Default actions._ Fail if there are any existing references to the key being updated or deleted, but allow this check to be deferred until the end of the transaction.
> ```

As of right now, NO ACTION And RESTRICT are the same.  Until we implement deferrable, they will always behave like Restrict.

And it's not immediately fail, it's fail at the end of a statement instead of the end of a transaction.

---

*[v2.0/foreign-key.md, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47e_SAtcle8OtD4En1:-L47e_SAtcle8OtD4En2:b-kv8pbi) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/foreign-key.md#L106)):*
> ```Markdown
>   );
> ~~~
> {{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it drops, so it should be used cautiously.{{site.data.alerts.end}}
> ```

Again, this should be expanded.

---

*[v2.0/update.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47eiuHYg36IBnsWhO-:-L47eiuHYg36IBnsWhO0:b-kr1kjs) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/update.md#L9)):*
> ```Markdown
> The `UPDATE` [statement](sql-statements.html) updates rows in a table.
> 
> {{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it updates, so it should be used cautiously. If you update a row that is referenced by a <a href="foreign-key.html">foreign key</a> and <code>ON UPDATE CASCADE</code> is set, all of the dependent rows will also be updated.{{site.data.alerts.end}}
> ```

and has an ON UPDATE foreign key action

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 3 of 5 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: 4 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[v2.0/foreign-key.md, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47dDxl7WvwSxuMF_K4:-L47fIm8iz6eD0jHnSBJ:b-oihat4) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/foreign-key.md#L57)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

As of right now, NO ACTION And RESTRICT are the same.  Until we implement deferrable, they will always behave like Restrict.

And it's not immediately fail, it's fail at the end of a statement instead of the end of a transaction.
</blockquote></details>

So maybe just put RESTRICT in as an alias of NO ACTION.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Review status: 4 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[v2.0/add-constraint.md, line 107 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47cpjiKdJ19KMHqlXB:-L47wTYwS610nqX9BE05:b-ngx2yt) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L107)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

There are two bullets here.  Did you mean to not have the first one?
</blockquote></details>

Typo - edited.

---

*[v2.0/add-constraint.md, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47cx1ds0b2DGZBX4IP:-L47wZgNd5XMtMOwr7qd:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L112)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

change this to "does not list the objects that it drops or updates"
</blockquote></details>

Done.

---

*[v2.0/add-constraint.md, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47d6BNn7cnKfnqiPuX:-L47wvIY1K-pB51jXQI5:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/add-constraint.md#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Add a link here to Foreign Key Actions?
</blockquote></details>

Done.

---

*[v2.0/create-table.md, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47f1nYUT8fgX_Y98Hm:-L48EvnqWcVbrSCS4SKY:ba3dfob) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/create-table.md#L329)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I don't know if this advice is required.  By default we are NO ACTION which is a restrict right now.
</blockquote></details>

Removed

---

*[v2.0/delete.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47ep3VqRNFYNcyCKZ3:-L48FM9NbiM3kay096rW:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/delete.md#L9)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Expand this for the other actions.
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47dDxl7WvwSxuMF_K4:-L48FkIlTB6KaDFtV-aP:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/foreign-key.md#L57)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

So maybe just put RESTRICT in as an alias of NO ACTION.
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47e_SAtcle8OtD4En1:-L48FqJnONdWmOKBFjLX:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/foreign-key.md#L106)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Again, this should be expanded.
</blockquote></details>

Done.

---

*[v2.0/update.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L47eiuHYg36IBnsWhO-:-L48I0lgarwFdMN7e8Dc:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/24b68dde6118ab64745070ad449d81aa044268a1/v2.0/update.md#L9)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

and has an ON UPDATE foreign key action
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f49e2dac63499722d4bb78e76de7516a2ec1a27b/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/c39ac23683a04b85e1663c1a196ab84fa5ce2611/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/700cc72f252c1fe041f9547dfe7407dd5c08a56f/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f6ed08ffc1e39bb588db3276683e12607f9ad172/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7b3549ffe11c0b275e14be8903907501eeb11cad/

Amruta-Ranade commented on v2.0/delete.md:
```diff
@@ -6,6 +6,8 @@ toc: false
 
 The `DELETE` [statement](sql-statements.html) deletes rows from a table.
 
+{{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it drops or updates, so it should be used cautiously. If you delete a row that is referenced by a <a href="foreign-key.html">foreign key</a> and <code>ON DELETE CASCADE</code> is set, all of the dependent rows will also be deleted.{{site.data.alerts.end}}
```
This seems abrupt to me - as a user, I would expect to first find information about how to use `DELETE` in normal scenarios, and then be warned about potential problems. Can this warning be moved to another part of the document, may be with an example of using `CASCADE`?

Amruta-Ranade commented on v2.0/foreign-key.md:
```diff
@@ -49,6 +49,18 @@ For example, if you have a Foreign Key constraint on columns `(A, B)` and try to
+----------|------------
+`ON DELETE NO ACTION` / `ON UPDATE NO ACTION` | _Default actions._ Fail if there are any existing references to the key being updated or deleted, but allow this check to be deferred until the end of the statement. <br><br>Aliases: `ON DELETE RESTRICT` / `ON UPDATE RESTRICT`
+`ON DELETE RESTRICT` / `ON UPDATE RESTRICT` | `RESTRICT` and `NO ACTION` are currently equivalent until options for deferring constraint checking are added. To set an existing foreign key action to `RESTRICT`, the foreign key constraint must be dropped and recreated.
+`ON DELETE CASCADE` / `ON UPDATE CASCADE` | When a referenced foreign key is deleted or updated, all rows referencing that key are deleted or updated. If there are other alterations to the row, such as a `SET NULL` or `SET DEFAULT`, the delete with take precedence. <br><br>Note that `CASCADE` does not list objects it drops or updates, so it should be used cautiously.
```
"the delete with take precedence" seems incomplete or missing code formatting.

lhirata replied:
Typo - fixed!

Amruta-Ranade commented on v2.0/foreign-key.md:
```diff
@@ -49,6 +49,18 @@ For example, if you have a Foreign Key constraint on columns `(A, B)` and try to
+
+Parameter | Description
+----------|------------
+`ON DELETE NO ACTION` / `ON UPDATE NO ACTION` | _Default actions._ Fail if there are any existing references to the key being updated or deleted, but allow this check to be deferred until the end of the statement. <br><br>Aliases: `ON DELETE RESTRICT` / `ON UPDATE RESTRICT`
```
In this table, I am unsure which actions I (the user) need to take and which actions CockroachDB performs that I am being informed about. For instance, "allow this check to be deferred" seems like something CockroachDB does (although I might be wrong), "the foreign key must be dropped and recreated" seems like something I would need to do (again, I might be wrong). Can we specify "you need to" or "CockroachDB does..." wherever possible?

lhirata replied:
Edited language to be clearer.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2385babfc7b838a98d998fef5a2dc4c1d8ac9bc8/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/fa619586f078ca33770c773a7d52c78072075559/

jseldess commented:
@lhirata, it looks like the minor fixes I made in https://github.com/cockroachdb/docs/pull/2449 didn't make it into this PR. Can you reapply those fixes? It's mostly code-block formatting, but there was one error in an example. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-:-L4bHoivXY79gMQ6L1XB:b-dbpx42)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Done.

---

Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-:-L4bJSkx8J-n3Wh9626B:b-896fix)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/dcb79bd16005c551dd8e5df3089a34216b7a2659/

jseldess commented:
*[v2.0/foreign-key.md, line 53 at r10](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L4bLAjgErIORL5POrw4:-L4bLAjgErIORL5POrw5:bnujxpr) ([raw file](https://github.com/cockroachdb/docs/blob/fa619586f078ca33770c773a7d52c78072075559/v2.0/foreign-key.md#L53)):*
I think it'd be helpful to provide more of an intro here, for example:

> When you set a foreign key constraint, you can control what happens to the constrained column when the column its referencing (the foreign key) is deleted or updated. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@lhirata, unfortunately, although we talked about making updates to all these pages, I'm afraid the overall sense of the feature is getting lost (on me, at least). Now that you've got this first iteration, I think we need to step back and rethink where and how we explain this. Let's set up a meeting.  

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-:-L4bOfgxO6CYXKQLrDh9:b-2nf6az)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/c122a5cec24ff253a54618a139aef017dbed0789/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9643792fff75ba5fd4639110d91c93138f9e2873/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/529a6a4f4816370aaf293e8fbc95b9a8393e5403/

lhirata commented:
@BramGruneir & @jseldess, can you take another look? I cleaned up the examples on CREATE TABLE and Foreign Key.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b92a14dde814d85d221a28f12bbf84e953d1a4a4/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f1c54fa482ce71bd4b4fd22b36e1b25cc5e12d31/

jseldess commented on v2.0/foreign-key.md:
```diff
@@ -49,6 +49,17 @@ For example, if you have a Foreign Key constraint on columns `(A, B)` and try to
 
 However, allowing _NULL_ values in either your foreign key or referenced columns can degrade their referential integrity. To avoid this, you can use the [Not Null constraint](not-null.html) on both sets of columns when [creating your tables](create-table.html). (The Not Null constraint cannot be added to existing tables.)
 
+### Foreign Key Actions
```
Let's put the "new in" callout in the header instead so the newness shows up in the toc:

```
### Foreign Key Actions <span class="version-tag">New in v2.0</span>
```

lhirata replied:
Done / Updated links

jseldess commented on v2.0/foreign-key.md:
```diff
@@ -119,25 +131,48 @@ CREATE TABLE packages (
 ~~~
 
-## Usage Example
+## Usage Examples
```
These new examples are excellent!

lhirata replied:
Great!

jseldess commented on v2.0/foreign-key.md:
```diff
@@ -119,25 +131,48 @@ CREATE TABLE packages (
+
+In this example, we'll create a table with a foreign key constraint with the default [actions](#foreign-key-actions) (`ON UPDATE NO ACTION` / `ON DELETE NO ACTION`).
 
+First, create the referenced table.
```
nit: Whenever you are introducing a command, I'd suggest using a colon (this should eventually be in our style guide).

> First, create the referenced table:

lhirata replied:
Done

jseldess commented on v2.0/create-table.md:
```diff
@@ -205,45 +205,82 @@ We also have other resources on indexes:
 - Referenced columns must contain only unique values. This means the `REFERENCES` clause must use exactly the same columns as a [Primary Key](primary-key.html) or [Unique](unique.html) constraint.
 
-In this example, we'll show a series of tables using different formats of foreign keys.
+<span class="version-tag">New in v2.0:</span> You can include a [foreign key action](foreign-key.html#foreign-key-actions) to specify what happens when a column referenced by a foreign key is updated or deleted. The default actions are `ON UPDATE NO ACTION` and `ON DELETE NO ACTION`.
```
nit: `a column referenced by a foreign key is updated` > `a column referenced by a foreign key constraint is updated`

lhirata replied:
Done

jseldess commented on v2.0/create-table.md:
```diff
@@ -205,45 +205,82 @@ We also have other resources on indexes:
+In this example, we use `ON DELETE CASCADE` (i.e., when row referenced by a foreign key constraint is deleted, all dependent rows are also deleted).
 
-> CREATE TABLE products (sku STRING PRIMARY KEY, price DECIMAL(9,2));
+{{site.data.alerts.callout_danger}}<code>CASCADE</code> does not list objects it drops, so it should be used cautiously.{{site.data.alerts.end}}
```
I'd remove this. When mentioning that this example will use `CASCADE`, you already explain what that means.

lhirata replied:
Done

jseldess commented on v2.0/delete.md:
```diff
@@ -6,6 +6,8 @@ toc: false
 
 The `DELETE` [statement](sql-statements.html) deletes rows from a table.
 
+{{site.data.alerts.callout_danger}}If you delete a row that is referenced by a <a href="foreign-key.html">foreign key</a> and has an <a href="foreign-key.html#foreign-key-actions"><code>ON DELETE</code> action</a>, all of the dependent rows will also be deleted or updated.{{site.data.alerts.end}}
```
nit: `a foreign key and has an ON DELETE action` > `a foreign key constraint with an ON DELETE action`

lhirata replied:
Done

jseldess commented on v2.0/update.md:
```diff
@@ -6,9 +6,11 @@ toc: false
 
 The `UPDATE` [statement](sql-statements.html) updates rows in a table.
 
+{{site.data.alerts.callout_danger}}If you update a row that contains a column referenced by a <a href="foreign-key.html">foreign key</a> and has an <a href="foreign-key.html#foreign-key-actions"><code>ON UPDATE</code> action</a>, all of the dependent rows will also be updated.{{site.data.alerts.end}}
```
nit: `a foreign key and has an ON UPDATE action` > `a foreign key constraint with an ON UPDATE action`

lhirata replied:
Done

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ace6695586fe6bcb920bd9eb859d0cd635f79e82/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/07c2454bda0e92bd024f481b9206313c3ac1fb59/

BramGruneir commented:
Left some comments.  This is looking really good.

---

Reviewed 1 of 1 files at r9, 5 of 5 files at r12.
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[v2.0/drop-column.md, line 25 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59mk893SmBcJoX2uBz:-L59mk8AA5Ax29_GSWK-:b-bdbb0g) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/drop-column.md#L25)):*
> ```Markdown
> | `table_name` | The name of the table with the column you want to drop. |
> | `name` | The name of the column you want to drop. |
> | `CASCADE` | Drop the column even if objects (such as [views](views.html)) depend on it; drop the dependent objects, as well.<br><br>`CASCADE` does not list objects it drops, so should be used cautiously. However, `CASCADE` will not drop dependent indexes; you must use [`DROP INDEX`](drop-index.html). <br><br><span class="version-tag">New in v2.0:</span> `CASCADE` will drop a column with a foreign key constraint if it is the only column in the reference. |
> ```

This is not part of my changes.  Not sure if you want to include it in this PR.

---

*[v2.0/foreign-key.md, line 134 at r11](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L3UzTlm6LqSSb2KLE6H-r11-134:-L59onuQrxL8kc2zYxLr:bjkizgz) ([raw file](https://github.com/cockroachdb/docs/blob/f1c54fa482ce71bd4b4fd22b36e1b25cc5e12d31/v2.0/foreign-key.md#L134)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Great! 
</blockquote></details>

Is it worth adding in more examples?  I agree that they're great!

---

*[v2.0/foreign-key.md, line 53 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59mvbZl7PIk8PskE5M:-L59mvbZl7PIk8PskE5N:bhv830n) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L53)):*
> ```Markdown
> 
> ### Foreign Key Actions <span class="version-tag">New in v2.0</span>
> You can specify an action to take when a referenced column is updated or deleted.
> ```

You can set both.  So perhaps 'and/or' is more appropriate than just 'or'.

---

*[v2.0/foreign-key.md, line 57 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59n6btNQB-6N8wu8yh:-L59n6btNQB-6N8wu8yi:bvl0wpy) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L57)):*
> ```Markdown
> Parameter | Description
> ----------|------------
> `ON DELETE NO ACTION` / `ON UPDATE NO ACTION` | _Default actions._ If there are any existing references to the key being updated or deleted, the transaction will fail at the end of the statement. <br><br>Aliases: `ON DELETE RESTRICT` / `ON UPDATE RESTRICT`
> ```

This might need to be split in two.

ON DELETE NO ACTION means that the referenced row can't be deleted, but it could be updated depending on the update action.
ON UPDATE NO ACTION mean that the referenced row can't be updated but it could be deleted depending on the delete action.  

---

*[v2.0/foreign-key.md, line 60 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59o8_qDBQpGO_guP5L:-L59o8_qDBQpGO_guP5M:b-60pihu) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L60)):*
> ```Markdown
> `ON DELETE RESTRICT` / `ON UPDATE RESTRICT` | `RESTRICT` and `NO ACTION` are currently equivalent until options for deferring constraint checking are added. To set an existing foreign key action to `RESTRICT`, the foreign key constraint must be dropped and recreated.
> `ON DELETE CASCADE` / `ON UPDATE CASCADE` | When a referenced foreign key is deleted or updated, all rows referencing that key are deleted or updated. If there are other alterations to the row, such as a `SET NULL` or `SET DEFAULT`, the delete will take precedence. <br><br>Note that `CASCADE` does not list objects it drops or updates, so it should be used cautiously.
> `ON DELETE SET NULL` / `ON UPDATE SET NULL` | When a referenced foreign key is deleted or updated, the columns of all rows referencing that key will be set to `NULL`. The column must allow `NULL` or this update will fail.
> ```

Maybe add 'respectively' to these?

---

*[v2.0/foreign-key.md, line 138 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59oFmr8VngSs-b720B:-L59oFmr8VngSs-b720C:b-km7wme) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L138)):*
> ```Markdown
> ### Use a Foreign Key Constraint with Default Actions
> 
> In this example, we'll create a table with a foreign key constraint with the default [actions](#foreign-key-actions-new-in-v2-0) (`ON UPDATE NO ACTION` / `ON DELETE NO ACTION`).
> ```

why the `/`?

---

*[v2.0/foreign-key.md, line 198 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59oWn-Q9x2NX87Wdpo:-L59oWn-Q9x2NX87Wdpp:b-vx1sbq) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L198)):*
> ```Markdown
> 
> Similarly, the deletion returns an error because `id = 100` is referenced and the default [foreign key action](#foreign-key-actions-new-in-v2-0) is enabled (`ON DELETE NO ACTION`).
> ```

Might be nice to also include two other examples that can be updated and deleted because they are not referenced.

---

*[v2.0/foreign-key.md, line 201 at r12](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L59oeYn9niz4XMhpjIt:-L59oeYn9niz4XMhpjIu:b-6ocffh) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L201)):*
> ```Markdown
> ### Use a Foreign Key Constraint with `CASCADE`
> 
> In this example, we'll create a table with a foreign key constraint with the [foreign key action](#foreign-key-actions-new-in-v2-0) `ON DELETE CASCADE`.
> ```

You're adding both on delete and on updated cascade.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-:-L59oxR6Xsca1pWwUSIN:bdf6ibz)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[v2.0/drop-column.md, line 25 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59mk893SmBcJoX2uBz:-L59xzmd_lBIP0qWtFRt:b-ln3e7g) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/drop-column.md#L25)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is not part of my changes.  Not sure if you want to include it in this PR.
</blockquote></details>

I'm not sure if we have best practices around this. @jseldess, should I put this edit in its own PR?

---

*[v2.0/foreign-key.md, line 53 at r10](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L4bLAjgErIORL5POrw4:-L5A1PD6EZ6oLasq-msi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/fa619586f078ca33770c773a7d52c78072075559/v2.0/foreign-key.md#L53)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

I think it'd be helpful to provide more of an intro here, for example:

> When you set a foreign key constraint, you can control what happens to the constrained column when the column its referencing (the foreign key) is deleted or updated. 
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 134 at r11](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L3UzTlm6LqSSb2KLE6H-r11-134:-L5AF-QZ8uEMxtHSM6LG:b-79b42t) ([raw file](https://github.com/cockroachdb/docs/blob/f1c54fa482ce71bd4b4fd22b36e1b25cc5e12d31/v2.0/foreign-key.md#L134)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is it worth adding in more examples?  I agree that they're great!
</blockquote></details>

Added to the "default actions" example, like you suggested.  I think that's sufficient!

---

*[v2.0/foreign-key.md, line 53 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59mvbZl7PIk8PskE5M:-L59z5Mm5UyT678IYoUy:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L53)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

You can set both.  So perhaps 'and/or' is more appropriate than just 'or'.
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 57 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59n6btNQB-6N8wu8yh:-L59zkzmVww6iEQ7Fbn_:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L57)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This might need to be split in two.

ON DELETE NO ACTION means that the referenced row can't be deleted, but it could be updated depending on the update action.
ON UPDATE NO ACTION mean that the referenced row can't be updated but it could be deleted depending on the delete action.  
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 60 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59o8_qDBQpGO_guP5L:-L5A-ss8XG3_3T52RT7-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L60)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Maybe add 'respectively' to these?
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 138 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59oFmr8VngSs-b720B:-L59yzXGDxhkOOtHcuch:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L138)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

why the `/`?
</blockquote></details>

Done.

---

*[v2.0/foreign-key.md, line 198 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59oWn-Q9x2NX87Wdpo:-L5AEsJk-1MQtJMV7NkD:bqmjwcz) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L198)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Might be nice to also include two other examples that can be updated and deleted because they are not referenced.
</blockquote></details>

Added.

---

*[v2.0/foreign-key.md, line 201 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59oeYn9niz4XMhpjIt:-L59yg7dD6-vxiJG3nsd:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/foreign-key.md#L201)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

You're adding both on delete and on updated cascade.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/0890c3e6bf05b2802f3739e92cd10b9f4f78b27c/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8034441472144c218d9c0f871ba13b32c3c32874/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/867ca8b5f30cc70de7a6ad3c14ac1c360f08c8c4/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/1bf7c16cf6e0853a1b44d0744a97978cd81a7290/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/30b1f131ebbd580fb06bc4a46143315d656aa882/

lhirata commented:
Review status: 4 of 6 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[v2.0/drop-column.md, line 25 at r12](https://beta.reviewable.io/reviews/cockroachdb/docs/2379#-L59mk893SmBcJoX2uBz:-L5AJk7-7anZUvYJMiLA:b-58ujei) ([raw file](https://github.com/cockroachdb/docs/blob/07c2454bda0e92bd024f481b9206313c3ac1fb59/v2.0/drop-column.md#L25)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

I'm not sure if we have best practices around this. @jseldess, should I put this edit in its own PR?
</blockquote></details>

Just pulled out this edit. Will open a new PR.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/61b84b0f9f084d5003fed8147810a155cc907364/

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-:-L5Asl1zOZ9cjSVhbO34:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[v2.0/foreign-key.md, line 134 at r11](https://reviewable.io:443/reviews/cockroachdb/docs/2379#-L3UzTlm6LqSSb2KLE6H-r11-134:-L5AsdukH5_7S8IsGiOH:b-rmu4wm) ([raw file](https://github.com/cockroachdb/docs/blob/f1c54fa482ce71bd4b4fd22b36e1b25cc5e12d31/v2.0/foreign-key.md#L134)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Added to the "default actions" example, like you suggested.  I think that's sufficient!
</blockquote></details>

I think it might be worthwhile to show how SET NULL and SET DEFAULT work too.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2379)*
<!-- Sent from Reviewable.io -->

lhirata commented:
@BramGruneir - will open a new issue for the SET NULL and SET DEFAULT examples
